// @svta/common-media-library/dist/cjs/webvtt.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebVttTransformStream = exports.WebVttTransformer = exports.WebVttParsingError = exports.WebVttParser = exports.toVttRegion = exports.toVttCue = exports.parseWebVtt = exports.createWebVttRegion = exports.createWebVttCue = void 0;
/**
 * @groupDescription WebVTT
 * A collection of tools for working with Web VTT.
 *
 * @see {@link https://www.w3.org/TR/webvtt1/ | WebVTT: The Web Video Text Tracks Format}
 *
 * @packageDocumentation
 */
var createWebVttCue_js_1 = require("./webvtt/createWebVttCue.js");
Object.defineProperty(exports, "createWebVttCue", { enumerable: true, get: function () { return createWebVttCue_js_1.createWebVttCue; } });
var createWebVttRegion_js_1 = require("./webvtt/createWebVttRegion.js");
Object.defineProperty(exports, "createWebVttRegion", { enumerable: true, get: function () { return createWebVttRegion_js_1.createWebVttRegion; } });
var parseWebVtt_js_1 = require("./webvtt/parseWebVtt.js");
Object.defineProperty(exports, "parseWebVtt", { enumerable: true, get: function () { return parseWebVtt_js_1.parseWebVtt; } });
var toVttCue_js_1 = require("./webvtt/toVttCue.js");
Object.defineProperty(exports, "toVttCue", { enumerable: true, get: function () { return toVttCue_js_1.toVttCue; } });
var toVttRegion_js_1 = require("./webvtt/toVttRegion.js");
Object.defineProperty(exports, "toVttRegion", { enumerable: true, get: function () { return toVttRegion_js_1.toVttRegion; } });
var WebVttParser_js_1 = require("./webvtt/WebVttParser.js");
Object.defineProperty(exports, "WebVttParser", { enumerable: true, get: function () { return WebVttParser_js_1.WebVttParser; } });
var WebVttParsingError_js_1 = require("./webvtt/WebVttParsingError.js");
Object.defineProperty(exports, "WebVttParsingError", { enumerable: true, get: function () { return WebVttParsingError_js_1.WebVttParsingError; } });
var WebVttTransformer_js_1 = require("./webvtt/WebVttTransformer.js");
Object.defineProperty(exports, "WebVttTransformer", { enumerable: true, get: function () { return WebVttTransformer_js_1.WebVttTransformer; } });
var WebVttTransformStream_js_1 = require("./webvtt/WebVttTransformStream.js");
Object.defineProperty(exports, "WebVttTransformStream", { enumerable: true, get: function () { return WebVttTransformStream_js_1.WebVttTransformStream; } });
