// @svta/common-media-library/dist/cjs/cmcd/CMCD_HEADER_MAP.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CMCD_HEADER_MAP = void 0;
const CMCD_OBJECT_js_1 = require("./CMCD_OBJECT.js");
const CMCD_REQUEST_js_1 = require("./CMCD_REQUEST.js");
const CMCD_SESSION_js_1 = require("./CMCD_SESSION.js");
const CMCD_STATUS_js_1 = require("./CMCD_STATUS.js");
/**
 * The map of CMCD keys to their appropriate header shard.
 *
 * @group CMCD
 *
 * @internal
 */
exports.CMCD_HEADER_MAP = {
    // Object
    br: CMCD_OBJECT_js_1.CMCD_OBJECT,
    ab: CMCD_OBJECT_js_1.CMCD_OBJECT,
    d: CMCD_OBJECT_js_1.CMCD_OBJECT,
    ot: CMCD_OBJECT_js_1.CMCD_OBJECT,
    tb: CMCD_OBJECT_js_1.CMCD_OBJECT,
    tpb: CMCD_OBJECT_js_1.CMCD_OBJECT,
    lb: CMCD_OBJECT_js_1.CMCD_OBJECT,
    tab: CMCD_OBJECT_js_1.CMCD_OBJECT,
    lab: CMCD_OBJECT_js_1.CMCD_OBJECT,
    url: CMCD_OBJECT_js_1.CMCD_OBJECT,
    // Request
    pb: CMCD_REQUEST_js_1.CMCD_REQUEST,
    bl: CMCD_REQUEST_js_1.CMCD_REQUEST,
    tbl: CMCD_REQUEST_js_1.CMCD_REQUEST,
    dl: CMCD_REQUEST_js_1.CMCD_REQUEST,
    ltc: CMCD_REQUEST_js_1.CMCD_REQUEST,
    mtp: CMCD_REQUEST_js_1.CMCD_REQUEST,
    nor: CMCD_REQUEST_js_1.CMCD_REQUEST,
    nrr: CMCD_REQUEST_js_1.CMCD_REQUEST,
    rc: CMCD_REQUEST_js_1.CMCD_REQUEST,
    sn: CMCD_REQUEST_js_1.CMCD_REQUEST,
    sta: CMCD_REQUEST_js_1.CMCD_REQUEST,
    su: CMCD_REQUEST_js_1.CMCD_REQUEST,
    ttfb: CMCD_REQUEST_js_1.CMCD_REQUEST,
    ttfbb: CMCD_REQUEST_js_1.CMCD_REQUEST,
    ttlb: CMCD_REQUEST_js_1.CMCD_REQUEST,
    cmsdd: CMCD_REQUEST_js_1.CMCD_REQUEST,
    cmsds: CMCD_REQUEST_js_1.CMCD_REQUEST,
    smrt: CMCD_REQUEST_js_1.CMCD_REQUEST,
    df: CMCD_REQUEST_js_1.CMCD_REQUEST,
    cs: CMCD_REQUEST_js_1.CMCD_REQUEST,
    // TODO: Which header to put the `ts` field is not defined yet.
    ts: CMCD_REQUEST_js_1.CMCD_REQUEST,
    // Session
    cid: CMCD_SESSION_js_1.CMCD_SESSION,
    pr: CMCD_SESSION_js_1.CMCD_SESSION,
    sf: CMCD_SESSION_js_1.CMCD_SESSION,
    sid: CMCD_SESSION_js_1.CMCD_SESSION,
    st: CMCD_SESSION_js_1.CMCD_SESSION,
    v: CMCD_SESSION_js_1.CMCD_SESSION,
    msd: CMCD_SESSION_js_1.CMCD_SESSION,
    // Status
    bs: CMCD_STATUS_js_1.CMCD_STATUS,
    bsd: CMCD_STATUS_js_1.CMCD_STATUS,
    cdn: CMCD_STATUS_js_1.CMCD_STATUS,
    rtp: CMCD_STATUS_js_1.CMCD_STATUS,
    bg: CMCD_STATUS_js_1.CMCD_STATUS,
    pt: CMCD_STATUS_js_1.CMCD_STATUS,
    ec: CMCD_STATUS_js_1.CMCD_STATUS,
    e: CMCD_STATUS_js_1.CMCD_STATUS,
};
