// node-datachannel/polyfill/index.js
import RTCCertificate from './RTCCertificate.js';
import RTCDataChannel from './RTCDataChannel.js';
import RTCDtlsTransport from './RTCDtlsTransport.js';
import RTCIceCandidate from './RTCIceCandidate.js';
import RTCIceTransport from './RTCIceTransport.js';
import RTCPeerConnection from './RTCPeerConnection.js';
import RTCSctpTransport from './RTCSctpTransport.js';
import RTCSessionDescription from './RTCSessionDescription.js';
import { RTCDataChannelEvent, RTCPeerConnectionIceEvent } from './Events.js';
import RTCError from './RTCError.js';

export {
    RTCCertificate,
    RTCDataChannel,
    RTCDtlsTransport,
    RTCIceCandidate,
    RTCIceTransport,
    RTCPeerConnection,
    RTCSctpTransport,
    RTCSessionDescription,
    RTCDataChannelEvent,
    RTCPeerConnectionIceEvent,
    RTCError,
};

export default {
    RTCCertificate,
    RTCDataChannel,
    RTCDtlsTransport,
    RTCIceCandidate,
    RTCIceTransport,
    RTCPeerConnection,
    RTCSctpTransport,
    RTCSessionDescription,
    RTCDataChannelEvent,
    RTCPeerConnectionIceEvent,
    RTCError,
};
