// mediasoup-client/lib/handlers/ortc/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.addNackSupportForOpus = addNackSupportForOpus;
exports.addHeaderExtensionSupport = addHeaderExtensionSupport;
exports.getMsidStreamIdAndTrackId = getMsidStreamIdAndTrackId;
/**
 * This function adds RTCP NACK support for OPUS codec in given capabilities.
 */
function addNackSupportForOpus(rtpCapabilities) {
    for (const codec of rtpCapabilities.codecs ?? []) {
        if ((codec.mimeType.toLowerCase() === 'audio/opus' ||
            codec.mimeType.toLowerCase() === 'audio/multiopus') &&
            !codec.rtcpFeedback?.some(fb => fb.type === 'nack' && !fb.parameter)) {
            if (!codec.rtcpFeedback) {
                codec.rtcpFeedback = [];
            }
            codec.rtcpFeedback.push({ type: 'nack' });
        }
    }
}
/**
 * This function adds the given RTP header extension to given capabilities.
 */
function addHeaderExtensionSupport(rtpCapabilities, headerExtension) {
    if (rtpCapabilities.headerExtensions?.some(exten => exten.kind === headerExtension.kind && exten.uri === headerExtension.uri)) {
        return;
    }
    if (!rtpCapabilities.headerExtensions) {
        rtpCapabilities.headerExtensions = [];
    }
    const setPreferredIds = new Set(rtpCapabilities.headerExtensions
        .filter(exten => exten.uri !== headerExtension.uri)
        .map(exten => exten.preferredId));
    let preferredId = 1;
    while (setPreferredIds.has(preferredId)) {
        ++preferredId;
    }
    const newHeaderExtension = {
        kind: headerExtension.kind,
        uri: headerExtension.uri,
        preferredId,
        preferredEncrypt: false,
        direction: headerExtension.direction,
    };
    rtpCapabilities.headerExtensions.push(newHeaderExtension);
}
function getMsidStreamIdAndTrackId(msid) {
    if (!msid || typeof msid !== 'string') {
        return { msidStreamId: undefined, msidTrackId: undefined };
    }
    /**
     * `msidStreamId` must be an id or '-' (no stream).
     * `msidTrackId` is an optional id.
     */
    const [msidStreamId, msidTrackId] = msid.trim().split(/\s+/);
    if (!msidStreamId) {
        return { msidStreamId: undefined, msidTrackId: undefined };
    }
    return { msidStreamId, msidTrackId };
}
