// firebase-admin/lib/esm/firestore/index.js
import mod from "../../firestore/index.js";

export const BulkWriter = mod.BulkWriter;
export const BundleBuilder = mod.BundleBuilder;
export const CollectionGroup = mod.CollectionGroup;
export const CollectionReference = mod.CollectionReference;
export const DocumentReference = mod.DocumentReference;
export const DocumentSnapshot = mod.DocumentSnapshot;
export const FieldPath = mod.FieldPath;
export const FieldValue = mod.FieldValue;
export const Filter = mod.Filter;
export const Firestore = mod.Firestore;
export const GeoPoint = mod.GeoPoint;
export const GrpcStatus = mod.GrpcStatus;
export const Query = mod.Query;
export const QueryDocumentSnapshot = mod.QueryDocumentSnapshot;
export const QueryPartition = mod.QueryPartition;
export const QuerySnapshot = mod.QuerySnapshot;
export const Timestamp = mod.Timestamp;
export const Transaction = mod.Transaction;
export const WriteBatch = mod.WriteBatch;
export const WriteResult = mod.WriteResult;
export const getFirestore = mod.getFirestore;
export const initializeFirestore = mod.initializeFirestore;
export const setLogFunction = mod.setLogFunction;
export const v1 = mod.v1;
