// hybrid-chunk-store/index.js
import FSAccessChunkStore from 'fsa-chunk-store'
import IDBChunkStore from '@thaunknown/idb-chunk-store'
import MemoryChunkStore from 'memory-chunk-store'
import CacheChunkStore from 'cache-chunk-store'

const isChrome = !!((typeof globalThis !== 'undefined' && globalThis) || (typeof self !== 'undefined' && self) || (typeof window !== 'undefined' && window) || (typeof global !== 'undefined' && global)).chrome

const limit = isChrome ? Infinity : 2147483648 - 16777216 // 2GB - 16MB

const FSASupport = typeof navigator !== 'undefined' && navigator.storage?.getDirectory && FileSystemFileHandle?.prototype?.createWritable

const noop = () => {}

export default class HybridChunkStore {
  constructor (chunkLength, opts = {}) {
    this.chunkLength = Number(chunkLength)
    if (!this.chunkLength) throw new Error('First argument must be a chunk length')

    this.length = Number(opts.length) || Infinity

    this.limit = opts.limit || limit
    this.fallbackStore = null
    this.dataStore = null
    this.chunkCount = 0
    this.stores = []

    this._init(opts)
    if (this.dataStore) {
      if (opts.max > 0) this.dataStore = new CacheChunkStore(this.dataStore, { max: opts.max })
      this.stores.push(this.dataStore)
    }
    this.stores.push(this.fallbackStore)

    // check if chunk stores have asyncIterators

    // if (this.stores.every(store => !!store[Symbol.asyncIterator])) {
    //   this[Symbol.asyncIterator] = joinIterator(this.stores)
    // }
  }

  _init (opts) {
    if (opts.onlyMem || this.limit < this.chunkLength) {
      this.fallbackStore = new MemoryChunkStore(this.chunkLength, opts)
      return
    }
    const ChunkStore = FSASupport ? FSAccessChunkStore : IDBChunkStore
    if (this.limit >= this.length) {
      this.fallbackStore = new ChunkStore(this.chunkLength, opts)
      return
    }

    this.chunkCount = Math.floor(Math.min(this.length, this.limit) / this.chunkLength)
    const length = this.chunkCount * this.chunkLength
    const remaining = this.length - length
    this.dataStore = new ChunkStore(this.chunkLength, { ...opts, length })
    this.fallbackStore = new MemoryChunkStore(this.chunkLength, { ...opts, length: remaining })
  }

  get (index, opts, cb) {
    if (index >= this.chunkCount) {
      this.fallbackStore.get(index - this.chunkCount, opts, cb)
    } else {
      this.dataStore.get(index, opts, cb)
    }
  }

  put (index, buf, cb) {
    if (index >= this.chunkCount) {
      this.fallbackStore.put(index - this.chunkCount, buf, cb)
    } else {
      this.dataStore.put(index, buf, cb)
    }
  }

  close (cb = noop) {
    Promise.all(this.stores.map(store => new Promise(resolve => store.close(resolve)))).then(values => {
      const err = values.find(value => value)
      cb(err)
    })
  }

  destroy (cb = noop) {
    Promise.all(this.stores.map(store => new Promise(resolve => store.destroy(resolve)))).then(values => {
      const err = values.find(value => value)
      cb(err)
    })
  }
}
